satellite/gc/gc_test.go

```diff
@@ -0,0 +1,126 @@
+
+package gc_test
+
+import (
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	"go.uber.org/zap"
+
+	"storj.io/storj/internal/memory"
+	"storj.io/storj/internal/testcontext"
+	"storj.io/storj/internal/testplanet"
+	"storj.io/storj/internal/testrand"
+	"storj.io/storj/pkg/encryption"
+	"storj.io/storj/pkg/paths"
+	"storj.io/storj/pkg/pb"
+	"storj.io/storj/pkg/storj"
+	"storj.io/storj/satellite"
+)
+
+// TestGarbageCollection does the following:
+// * Set up a network with one storagenode
+// * Upload two objects
+// * Delete one object from the metainfo service on the satellite
+// * Trigger a bloom filter generation
+// * Check that pieces of the deleted object are deleted on the storagenode
+// * Check that pieces of the kept object are not deleted on the storagenode
+func TestGarbageCollection(t *testing.T) {
+	testplanet.Run(t, testplanet.Config{
+		SatelliteCount: 1, StorageNodeCount: 1, UplinkCount: 1,
+		Reconfigure: testplanet.Reconfigure{
+			Satellite: func(log *zap.Logger, index int, config *satellite.Config) {
+				config.GarbageCollection.FalsePositiveRate = 0.0001
+				config.GarbageCollection.Interval = 0
+			},
+		},
+	}, func(t *testing.T, ctx *testcontext.Context, planet *testplanet.Planet) {
+		satellite := planet.Satellites[0]
+		upl := planet.Uplinks[0]
+		targetNode := planet.StorageNodes[0]
+
+		checker := satellite.Repair.Checker
+		checker.Loop.Stop()
+
+		// Upload two objects
+		testData1 := testrand.Bytes(8 * memory.KiB)
+		testData2 := testrand.Bytes(8 * memory.KiB)
+
+		err := upl.Upload(ctx, satellite, "testbucket", "test/path/1", testData1)
+		require.NoError(t, err)
+		deletedEncPath, pointerToDelete := getPointer(ctx, t, satellite, upl, "testbucket", "test/path/1")
+		var deletedPieceID storj.PieceID
+		for _, p := range pointerToDelete.GetRemote().GetRemotePieces() {
+			if p.NodeId == targetNode.ID() {
+				deletedPieceID = pointerToDelete.GetRemote().RootPieceId.Derive(p.NodeId, p.PieceNum)
+				break
+			}
+		}
+		require.NotNil(t, deletedPieceID)
+
+		err = upl.Upload(ctx, satellite, "testbucket", "test/path/2", testData2)
+		require.NoError(t, err)
+		_, pointerToKeep := getPointer(ctx, t, satellite, upl, "testbucket", "test/path/2")
+		var keptPieceID storj.PieceID
+		for _, p := range pointerToKeep.GetRemote().GetRemotePieces() {
+			if p.NodeId == targetNode.ID() {
+				keptPieceID = pointerToKeep.GetRemote().RootPieceId.Derive(p.NodeId, p.PieceNum)
+				break
+			}
+		}
+		require.NotNil(t, keptPieceID)
+
+		// Delete object from metainfo service on satellite
+		err = satellite.Metainfo.Service.Delete(ctx, deletedEncPath)
+		require.NoError(t, err)
+
+		// Check that piece of the deleted object is on the storagenode
+		pieceInfo, err := targetNode.DB.PieceInfo().Get(ctx, satellite.ID(), deletedPieceID)
+		require.NoError(t, err)
+		require.NotNil(t, pieceInfo)
+
+		// Sleep is here because we need at least one full second between the next piece tracker
```

Comment:
_Why_ do we need a second?  Sleep almost always means we're doing it wrong.

Reply:
In the `pieceInfo.GetPieceIDs` query, we are converting the piece creation timestamp as well as the filter creation timestamp to `datetime` in sql. This chops off all precision beyond seconds. In our test, the amount of time that elapses between piece uploads and the garbage collection loop is less than a second, meaning `datetime(piece_creation) < datetime(filter_creation)` is false unless we sleep for a second. We went with a sleep because any workaround would have been a lot of effort, and it would be very specific to this test

Reply:
That's a great reason!  I would include a least a little bit of this info in the comment.